server/src/main/java/org/elasticsearch/lucene/util/MatchAllBitSet.java

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.lucene.util;
+
+import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.util.BitSet;
+import org.apache.lucene.util.FixedBitSet;
+import org.apache.lucene.util.RamUsageEstimator;
+
+import java.io.IOException;
+
+/**
+ * An optimized implementation of {@link BitSet} that matches all documents to reduce memory usage.
+ */
+public final class MatchAllBitSet extends BitSet {
+    private static final long RAM_BYTES_USED = RamUsageEstimator.shallowSizeOfInstance(MatchAllBitSet.class);
+
+    private FixedBitSet bits;
+    private final int numBits;
+
+    public MatchAllBitSet(int numBits) {
+        this.numBits = numBits;
+    }
+
+    @Override
+    public void set(int i) {
+        if (bits != null) {
+            bits.set(i);
+        }
+    }
+
+    @Override
+    public boolean getAndSet(int i) {
+        if (bits != null) {
+            return bits.getAndSet(i);
+        } else {
+            return true;
+        }
+    }
+
+    private void initializeBitSet() {
+        bits = new FixedBitSet(numBits);
+        bits.set(0, bits.length());
+    }
+
+    @Override
+    public void clear(int i) {
+        if (bits == null) {
+            initializeBitSet();
+        }
+        bits.clear(i);
```

Comment:
Do we actually need to support clearing bits on this `MatchAllBitSet`? It's counter intuitive to me that something called `MatchAllBitSet` could have some of its bits not set.

Reply:
Good call. I've pushed https://github.com/elastic/elasticsearch/commit/0f12598c31a46579677a277341c08c2c98cbc3da to make it read-only.